### PR TITLE
Adapt setuptools_scm so we can build the package from the GitHub tarball

### PR DESCRIPTION
**Pull request by ocefpaf**
@huard this should make it easier to distribute xESMF outside of PyPI. Before, the GitHub tarball did not contain the version metadata b/c `setuptools_scm` works with either a frozen sdist or a git clone. (The GitHub tarball is neither.)

Now that metadata is propagated via the `setuptools_scm_git_archive` plugin, so people can do a GitHub release install and get the correct version. I'll explain the details inline in the changes below.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,3 +1,7 @@
+[build-system]
```
Comment (ocefpaf) [inline pyproject.toml:1]: This also makes the package pep517/518 compliant by declaring the build system and build dependencies. Modern pip "knows" to install this and call the right build backend.

File: setup.py [hunk 2]
```diff
@@ -9,8 +9,11 @@
 URL = 'https://github.com/pangeo-data/xESMF'
 LICENSE = 'MIT'
 PYTHON_REQUIRES = '>=3.6'
-USE_SCM_VERSION = {'version_scheme': 'post-release', 'local_scheme': 'dirty-tag'}
-SETUP_REQUIRES = ['setuptools_scm', 'setuptools>=30.3.0']
+USE_SCM_VERSION = {
+    "write_to": "xesmf/_version.py",
```
Comment (ocefpaf) [inline setup.py:13]: The version file is much faster than `pkg_resources` b/c, which can get really slow depending on the packages installed on your env.

File: setup.py [hunk 3]
```diff
@@ -59,5 +62,4 @@ def readme():
     url=URL,
     packages=find_packages(),
     use_scm_version=USE_SCM_VERSION,
-    setup_requires=SETUP_REQUIRES,
```
Comment (ocefpaf) [inline setup.py:62]: This is handled in the pyproject.toml now. We can move pretty much everything from the setup.py to the toml and cfg files.

File: xesmf/__init__.py [hunk 4]
```diff
@@ -1,10 +1,7 @@
...
 from . import data, util
 from .frontend import Regridder, SpatialAverager
 
 try:
-    __version__ = get_distribution(__name__).version
-except DistributionNotFound:  # pragma: no cover
-    __version__ = '0.0.0'  # pragma: no cover
+    from ._version import __version__
+except ImportError:
+    __version__ = "unknown"
```
Comment (ocefpaf) [inline xesmf/__init__.py:7]: 0.0.0 gives the idea of a release. "unknown" is probably a better tag for something that is unreleased.

**Review by huard — APPROVED** [review]
I won't pretend I understand what's going on, but sounds good.

**Comment by andersy005** [comment]
Thank you for putting this together, @ocefpaf! I think we still need `# flake8: noqa` at the top of `__init__.py` in order to appease flake8.

File: setup.py [hunk 5]
```diff
@@ -19,8 +22,8 @@
 else:
     INSTALL_REQUIRES = [
         'esmpy>=8.0.0',
-        'xarray <= 0.16.0',
-        'numpy >=1.16',
+        'xarray<=0.16.0',
```
Comment (dcherian) [inline setup.py:25]:
```suggestion
        'xarray!=0.16.1',
```

shouldn't this be enough?

**Review by andersy005 — APPROVED** [review]

**Comment by ocefpaf** [comment]
> I won't pretend I understand what's going on, but sounds good.

@huard in theory, if we sticky to the same tagging scheme of `vMAJOR.MINOR.MICRO`, the conda-forge bots will find the releases automatically without the need for us to mint a new one for every release.

**Pull request merged**